rust_src/src/process.rs

```diff
@@ -273,14 +282,65 @@ pub fn set_process_buffer(process: LispObject, buffer: LispObject) -> LispObject
     }
     p_ref.set_buffer(buffer);
     let process_type = p_ref.process_type;
-    if process_type.eq(Qnetwork) || process_type.eq(Qserial) || process_type.eq(Qpipe) {
+    let netconn1_p = process_type.eq(Qnetwork);
+    let serialconn1_p = process_type.eq(Qserial);
+    let pipeconn1_p = process_type.eq(Qpipe);
+
+    if netconn1_p || serialconn1_p || pipeconn1_p {
         let childp = p_ref.childp;
         p_ref.set_childp(plist_put(childp, QCbuffer, buffer));
     }
     unsafe { setup_process_coding_systems(process.to_raw()) };
     buffer
 }
 
+/// Give PROCESS the filter function FILTER; nil means default.
+/// A value of t means stop accepting output from the process.
+///
+/// When a process has a non-default filter, its buffer is not used for output.
+/// Instead, each time it does output, the entire string of output is
+/// qpassed to the filter.
+///
+/// The filter gets two arguments: the process and the string of output.
+/// The string argument is normally a multibyte string, except:
+/// - if the process's input coding system is no-conversion or raw-text,
+///   it is a unibyte string (the non-converted input), or else
+/// - if `default-enable-multibyte-characters' is nil, it is a unibyte
+///   string (the result of converting the decoded input multibyte
+///   string to unibyte with `string-make-unibyte').
+#[lisp_fn]
+pub fn set_process_filter(process: LispObject, mut filter: LispObject) -> LispObject {
+    let mut p_ref = process.as_process_or_error();
+
+    // Don't signal an error if the process's input file descriptor
+    // is closed.  This could make debugging Lisp more difficult,
+    // for example when doing something like
+    //
+    // (setq process (start-process ...))
+    // (debug)
+    // (set-process-filter process ...)
+    if filter.is_nil() {
+        filter = Qinternal_default_process_filter;
+    }
+    unsafe { pset_filter(p_ref.as_mut(), filter) }
+
+    if p_ref.infd() >= 0 {
+        unsafe { set_process_filter_masks(p_ref.as_ptr()) };
+    }
+
```

Comment:
And now you have used them twice. Why not make them functions? They do not need to be public.